### PR TITLE
[Java] CWE-488: Exposure of Data Element to Wrong Session

### PR DESCRIPTION
**Pull request by haby0**
Store the value of the request parameter in a non-final member field. When the program does not correctly process the request value, it will cause the data to be provided to the wrong session or used by the wrong session. For example, in the following program, when user `A` The input value is `admin`, and the program outputs the `password` of `admin`. User `B` has no input value, and the program will output the `password` of user `A`.

```java

String name;

@Override
protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException {
    String tempName = req.getParameter("name");
    if (!StringUtils.isEmpty(tempName)) {
        name = tempName;
    }

    .... // Query the password corresponding to the user name
     
    PrintWriter pw = resp.getWriter();
    pw.println("user : " + name + "    password is : " + password);
}
```

### CODE REVIEW

**Comment by haby0** [comment]
@smowton Could you please help me review this pr?

File: java/ql/src/semmle/code/java/frameworks/Servlets.qll [hunk 1]
```diff
@@ -337,3 +337,16 @@ predicate isRequestGetParamMethod(MethodAccess ma) {
   ma.getMethod() instanceof ServletRequestGetParameterMapMethod or
   ma.getMethod() instanceof HttpServletRequestGetQueryStringMethod
 }
+
+/**
+ * The method is `doGet` or `doPost`.
+ */
+class ServletRequestMethod extends Method {
```
Comment (smowton) [inline java/ql/src/semmle/code/java/frameworks/Servlets.qll:344]: This nearly duplicates `isServletRequestMethod`; use that instead.

**Comment by smowton** [comment]
Passing this one to @owen-mc

File: java/ql/src/semmle/code/java/frameworks/Servlets.qll (hunk 1 above)
Comment (haby0) [reply java/ql/src/semmle/code/java/frameworks/Servlets.qll]: @smowton Thanks for review, I made a change here.

**Comment by haby0** [comment]
@owen-mc hi, Do you have any suggestions for this pr?

**Comment by owen-mc** [comment]
@haby0 I'm sorry I've taken so long to reply.

If I've understood `InputJudgeConfig` correctly, you're trying to say that if the taint flow from `source` to `sink` passes through a barrier guard that guarantees that the value is not empty then we should ignore it. (It doesn't quite do that because you don't make sure it's the same `sink` - there could be two `WrongSessionSink`s that `source` flows to.)

The good news is that there is already a way of doing this in the taint flow library: [isSanitizerGuard](https://github.com/github/codeql/blob/2172fb6e650880d68747f9204d50e7af1fefc624/java/ql/src/semmle/code/java/dataflow/internal/tainttracking1/TaintTrackingImpl.qll#L93). Unfortunately it doesn't currently have any uses in the java library for me to point you to. There are uses in other languages. Some of them have slightly different implementations to Java, but C# seems to be pretty much the same, so [here's an example from C#](https://github.com/github/codeql/blob/fddbce0b7bb8dc7e97971a0af7b01902954d1d49/csharp/ql/src/semmle/code/csharp/security/dataflow/UrlRedirect.qll#L110). You're already using `DataFlow::BarrierGuard` so it shouldn't need much changed. I'd make a class that extends `DataFlow::BarrierGuard` for each of the disjuncts in `isSink` in `InputJudgeConfig`. The characteristic predicate should specify the node that you currently assign to `bg` (`ma` or `nee`), and you should implement `checks(Expr e, boolean branch)` so what you currently call `sink.asExpr()` is `e` and `branch` is the return value of the method/the value of the `NEExpr` which means that taint flow should be blocked. Then you should override `isSanitizerGuard(DataFlow::BarrierGuard guard)` in `WrongSessionConfig` to say that `guard` is an instance of one of those three classes.

By the way, when I ran the tests locally they failed. An extra result was found 
```
+| WrongSessionServlet.java:21:19:21:25 | greetee | WrongSessionServlet.java:18:26:18:52 | getParameter(...) : String | WrongSessionServlet.java:21:19:21:25 | greetee | Wrong session might include code from $@. | WrongSessionServlet.java:18:26:18:52 | getParameter(...) | this user input |
```

**Comment by haby0** [comment]
> @haby0 I'm sorry I've taken so long to reply.
> 
> If I've understood `InputJudgeConfig` correctly, you're trying to say that if the taint flow from `source` to `sink` passes through a barrier guard that guarantees that the value is not empty then we should ignore it. (It doesn't quite do that because you don't make sure it's the same `sink` - there could be two `WrongSessionSink`s that `source` flows to.)
> 
> The good news is that there is already a way of doing this in the taint flow library: [isSanitizerGuard](https://github.com/github/codeql/blob/2172fb6e650880d68747f9204d50e7af1fefc624/java/ql/src/semmle/code/java/dataflow/internal/tainttracking1/TaintTrackingImpl.qll#L93). Unfortunately it doesn't currently have any uses in the java library for me to point you to. There are uses in other languages. Some of them have slightly different implementations to Java, but C# seems to be pretty much the same, so [here's an example from C#](https://github.com/github/codeql/blob/fddbce0b7bb8dc7e97971a0af7b01902954d1d49/csharp/ql/src/semmle/code/csharp/security/dataflow/UrlRedirect.qll#L110). You're already using `DataFlow::BarrierGuard` so it shouldn't need much changed. I'd make a class that extends `DataFlow::BarrierGuard` for each of the disjuncts in `isSink` in `InputJudgeConfig`. The characteristic predicate should specify the node that you currently assign to `bg` (`ma` or `nee`), and you should implement `checks(Expr e, boolean branch)` so what you currently call `sink.asExpr()` is `e` and `branch` is the return value of the method/the value of the `NEExpr` which means that taint flow should be blocked. Then you should override `isSanitizerGuard(DataFlow::BarrierGuard guard)` in `WrongSessionConfig` to say that `guard` is an instance of one of those three classes.
> 
> By the way, when I ran the tests locally they failed. An extra result was found
> 
> ```
> +| WrongSessionServlet.java:21:19:21:25 | greetee | WrongSessionServlet.java:18:26:18:52 | getParameter(...) : String | WrongSessionServlet.java:21:19:21:25 | greetee | Wrong session might include code from $@. | WrongSessionServlet.java:18:26:18:52 | getParameter(...) | this user input |
> ```

Thanks for the review, I need some time to deal with it.

**Comment by owen-mc** [comment]
Actually, I've realised that I misunderstood the intention of `InputJudgeConfig`. Please ignore my suggestions relating to it.

**Comment by haby0** [comment]
> Actually, I've realised that I misunderstood the intention of `InputJudgeConfig`. Please ignore my suggestions relating to it.

Received, I am also making some changes to the query.

**Review by wood14 — APPROVED** [review]

**Comment by owen-mc** [comment]
> Received, I am also making some changes to the query.

Okay, I will hold off from reviewing until you've updated it then.

**Comment by smowton** [comment]
Noting securitylab have closed their ticket; drop this?

**Comment by haby0** [comment]
> Noting securitylab have closed their ticket; drop this?

..., I am making some improvements to the query.

I turned off this pr first.

**Pull request closed without merging**